Add allcolumnsearch() function for searching across all text columns

pchandra345 opened the pull request:
## Summary
  Implements `allcolumnsearch()` function that searches for a term across all VARCHAR and CHAR columns in a query, similar to New Relic NRDB's allcolumnsearch() function.

  The function expands `allcolumnsearch('term')` into `(col1 LIKE '%term%' OR col2 LIKE '%term%' OR ...)` during query analysis, following Trino's pattern for row filter injection.

  ## Implementation Details
  - Table function implementation using AbstractConnectorTableFunction
  - Analysis phase: Validates search term, identifies VARCHAR/CHAR columns, validates at least one searchable column exists
  - Execution phase: Filters rows using Page.getPositions() for efficient page-level filtering
  - Regex support: Case-insensitive pattern matching with full regex support
  - Universal compatibility: Works across all Trino connectors (no connector-specific code)


  ## Features
  - Case-insensitive regex pattern matching
  - Efficient page-level row filtering
  - Handles NULL values correctly
  - Works with all query contexts: JOINs, subqueries, WHERE, ORDER BY, LIMIT
  - Comprehensive error handling (null/empty search term, invalid regex, no string columns)
  - Cross-connector support (TPCH, PostgreSQL, Iceberg, etc.)
  - Fully qualified name support (system.builtin.allcolumnsearch)


  ## Files Changed
  - core/trino-main/src/main/java/io/trino/operator/table/AllColumnSearchFunction.java - Table function implementation
  - core/trino-main/src/main/java/io/trino/connector/system/SystemConnectorModule.java - Function registration
  - core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java - Processor provider mapping
  - testing/trino-tests/src/test/java/io/trino/tests/TestAllColumnSearchFunction.java - Comprehensive test suite (19 test cases)

onkar-r commented on core/trino-main/src/main/java/io/trino/operator/scalar/AllColumnSearchFunction.java:
```diff
@@ -0,0 +1,38 @@
+
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+
+public final class AllColumnSearchFunction
```
It'd be good to provide a brief summary and an example usage of this function.

pchandra345 replied:
Update the docs

ebyhr commented:
@pchandra345 I recommend getting consensus on this syntax before doing further work. I'm not sure we really want this function.

pchandra345 commented:
> @pchandra345 I recommend getting consensus on this syntax before doing further work. I'm not sure we really want this function.

Please take a look into this thread:- (https://trinodb.slack.com/archives/CP1MUNEUX/p1765463946141919)

pchandra345 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -5066,6 +5069,9 @@ private void analyzeSelectSingleColumn(
         private void analyzeWhere(Node node, Scope scope, Expression predicate)
         {
+            // Expand allcolumnsearch() BEFORE analysis so expanded AST gets properly analyzed
+            predicate = expandAllColumnSearch(predicate, scope);
```
- Hidden/meta columns: My current implementation includes all visible columns. Should we explicitly exclude system columns?
  - Non-WHERE contexts: I added validations to prevent use in ORDER BY, JOIN ON  and select clause specifically because the semantics are unclear there.

pchandra345 replied:
Please take a look into this thread:- (https://trinodb.slack.com/archives/CP1MUNEUX/p1765463946141919)

pchandra345 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -6185,4 +6206,181 @@ public List<TableArgumentAnalysis> getTableArgumentAnalyses()
+    private static class AllColumnSearchValidator
+            extends DefaultExpressionTraversalVisitor<Void>
+    {
+        private final String context;
```
the context field at line is used in the error message at line 6238 to provide users with information about which clause contained the invalid allcolumnsearch() usage.

pchandra345 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -6185,4 +6206,181 @@ public List<TableArgumentAnalysis> getTableArgumentAnalyses()
+        @Override
+        protected Void visitFunctionCall(FunctionCall node, Void context)
+        {
+            if (node.getName().getSuffix().equalsIgnoreCase("allcolumnsearch")) {
```
Updated

pchandra345 commented:
> > they're connector-specific
> 
> @pchandra345 Have you confirmed [ExcludeColumnsFunction](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/table/ExcludeColumnsFunction.java) in the engine?

 Thanks for bringing up ExcludeColumnsFunction! You're right that it's a built-in table function (registered with GlobalSystemConnector) that works across all connectors.

   ExcludeColumnsFunction works as a table function because it:
  - Performs column projection (schema manipulation), not row filtering
  - Naturally belongs in FROM clause (defining what table structure to query)
  - Passes through all rows unchanged (line 150: usedInputAndProduced)

  allcolumnsearch() performs row filtering based on content matching, which is semantically a WHERE clause operation.

  The analyzer-based expansion approach correctly implements WHERE clause semantics while maintaining universal cross-connector support and pushdown optimization potential.

Praveen2112 reviewed (commented):
Can we squash both the commits to one.

pchandra345 committed:
Add allcolumnsearch() table function with comprehensive tests

Implements a new table function that searches for a pattern across all
VARCHAR/CHAR columns in a table and returns matching rows.

Features:
- Searches across all string columns automatically
- Supports regex patterns
- Case-sensitive and case-insensitive modes (case_sensitive parameter)
- Validates input parameters and regex patterns
- Handles NULL values gracefully

Includes 39 comprehensive tests covering:
- Basic search functionality
- Case sensitivity modes
- Regex pattern matching
- Parameter validation (20 failure test cases)
- Type validation
- Edge cases (empty tables, NULL values, etc.)

All tests passing.

pchandra345 commented on core/trino-main/src/main/java/io/trino/operator/table/AllColumnSearchFunction.java:
```diff
@@ -0,0 +1,212 @@
+                List<Integer> searchableColumns = functionHandle.searchableColumnIndices();
+
+                // Compile regex pattern once for reuse
+                Pattern searchPattern = Pattern.compile(searchTerm, Pattern.CASE_INSENSITIVE);
```
Added
